src/main/java/com/arpnetworking/metrics/mad/PeriodWorker.java

```diff
@@ -92,40 +109,57 @@ public void preRestart(final Throwable reason, final Optional<Object> message) t
         return receiveBuilder()
                 .match(Record.class, this::processRecord)
                 .matchEquals(ROTATE_MESSAGE, m -> rotateAndSchedule())
+                .matchEquals(IDLE_CHECK_MESSAGE, m -> checkForIdle())
                 .build();
     }
 
     private void rotateAndSchedule() {
         final ZonedDateTime now = ZonedDateTime.now();
-        _hasRotateScheduled = false;
+        _nextScheduledRotationTime = Optional.empty();
         performRotation(now);
         scheduleRotation(now);
     }
 
+    private void checkForIdle() {
+        // The check for both no scheduled rotations and having received records
+        // is necessary to ensure that the idle message is only sent after at
+        // least the minimum time has elapsed and no future rotations exist. If
+        // this were only based on rotations then the idle trigger could fire
+        // after the current rotation but before data for the next rotation
+        // arrives, therefore not ensuring that the minimum time has elapsed.
+        if (!_nextScheduledRotationTime.isPresent() && !_hasReceivedRecords) {
+            _aggregator.tell(new Aggregator.PeriodWorkerIdle(_key), self());
+        }
+        _hasReceivedRecords = false;
+    }
+
     private void scheduleRotation(final ZonedDateTime now) {
-        final Optional<ZonedDateTime> rotateAt = getRotateAt();
+        if (_nextScheduledRotationTask != null) {
+            _nextScheduledRotationTask.cancel();
+            _nextScheduledRotationTask = null;
+        }
 
-        if (rotateAt.isPresent()) {
+        _nextScheduledRotationTime = getRotateAt();
+
+        if (_nextScheduledRotationTime.isPresent()) {
             // If we we don't need to wait then just set the scheduled delay to 0.
             // If we need to wait a really small amount of time, set the delay to a minimum to avoid sleep thrashing.
             // Otherwise schedule the next rotation at the predicted time.
             // Finally, if we wake-up and there's nothing to rotate we'll just re-apply these rules.
 
-            Duration timeToRotate = Duration.between(now, rotateAt.get());
+            Duration timeToRotate = Duration.between(now, _nextScheduledRotationTime.get());
             if (timeToRotate.isNegative()) {
                 timeToRotate = Duration.ZERO;
             } else if (timeToRotate.compareTo(MINIMUM_ROTATION_CHECK_INTERVAL) < 0) {
                 timeToRotate = MINIMUM_ROTATION_CHECK_INTERVAL;
             }
 
-            _nextScheduledRotation = context().system().scheduler().scheduleOnce(
+            _nextScheduledRotationTask = context().system().scheduler().scheduleOnce(
```

Comment:
Please use the timers() call to schedule a single-shot timer. This will also eliminate/simplify some of the clean-up code as the base class will handle canceling named timers.

Reply:
Done